engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/ReentrantLocking.scala

```diff
@@ -248,7 +248,10 @@ class ReentrantLocking(logger: TruffleLogger) extends Locking {
     contextMapLock.lock()
     try {
       if (contextLocks.contains(contextId)) {
-        contextLocks(contextId).unlock()
+        assertNotLocked(
```

Comment:
I don't understand the desired behavior. Why do we have `removeContextLock` at all? Because of [this line](https://github.com/enso-org/enso/blob/dee9e079d40b1df72731cc5dc7a44d1212fb4c68/engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/command/DestroyContextCmd.scala#L48)? Then just remove it!

Reply:
Yes. I suppose it was added to avoid creating new instances of `ReentrantLock`. Once the lock is created it is re-used. Might be an example of pre-mature optimization? It saves roughly 10ms per command (as that particular lock is required quite often).

Reply:
Everywhere, where `ctx.locking.withContextLock(` is used, there is an instance of `RuntimeContext`. To 

> avoid creating new instances of ReentrantLock

the `RuntimeContext` shall obtain some object instance via `ctx.locking.createContextLock` and then use it via `ctx.locking.withContextLock(` until it is needed. At the end it should _close it_ (`removeContextLock`).

Reply:
Possible API could look like:
```diff
diff --git engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java
index fc6e3f6389..b223ce34fb 100644
--- engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java
+++ engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/Locking.java
@@ -37,21 +37,25 @@ public interface Locking {
   <T> T withPendingEditsLock(Class<?> where, Callable<T> callable);
 
   /**
-   * Executes `callable` while holding a context lock
+   * Allocates new handle for use in {@link #withContextLock} associated with given context ID.
    *
-   * @param contextId context id for which the lock is being requested
-   * @param where the class requesting the lock
-   * @param callable code to be executed while holding the lock
-   * @return the result of calling `callable` or null, if no result is expected
+   * @param contextId context ID
+   * @return an object to pass to {@link #withContextLock} and to {@link AutoCloseable#close()} when
+   *     no longer needed
    */
-  <T> T withContextLock(UUID contextId, Class<?> where, Callable<T> callable);
+  AutoCloseable createContextLock(UUID contextId);
 
   /**
-   * Removes a context lock.
+   * Executes `callable` while holding a context lock
    *
-   * @param contextId a context to remove
+   * @param contextLockHandle handle created by {@link #createContextLock(UUID)}
+   * @param where the class requesting the lock
+   * @param callable code to be executed while holding the lock
+   * @return the result of calling `callable` or null, if no result is expected
+   * @throws ClassCastException if {@code contextLockHandle} isn't coming from {@link
+   *     #createContextLock(UUID)}
    */
-  void removeContextLock(UUID contextId);
+  <T> T withContextLock(AutoCloseable contextLockHandle, Class<?> where, Callable<T> callable);
 
   /**
    * Executes `callable` while holding a file lock
```